torch/csrc/jit/passes/reconstruct_scopes.cpp

```diff
@@ -7,77 +7,163 @@ namespace jit {
 
+  std::string getModuleName(const Module& module, const std::string& prefix);
   void constructFunctionToModuleMap(const Module& module);
   void constructRelativeNamesForModules(
       const Module& module,
       const std::string& prefix);
 
-  std::string getScopeString(Function* f) const;
+  std::string getScopeString(const InlinedCallStackEntry& frame) const;
+
+  void appendSourceRangeInfo(
+      std::string& scopeString,
+      const InlinedCallStackEntry& frame) const;
 };
 
 void ReconstructScopesPass::constructFunctionToModuleMap(const Module& module) {
   for (const auto& method : module.get_methods()) {
-    func_to_module[&method.function()] = module._ivalue();
+    Function* func_ptr = &method.function();
+    if (!class_types_are_not_unique_ &&
+        func_to_module.find(func_ptr) != func_to_module.end()) {
+      class_types_are_not_unique_ = true;
+    }
+    func_to_module[func_ptr] = module._ivalue();
   }
   for (const Module& m : module.children()) {
     constructFunctionToModuleMap(m);
   }
 }
 
+std::string ReconstructScopesPass::getModuleName(
+    const Module& module,
+    const std::string& prefix) {
+  std::string moduleType = module.type()->str();
+  size_t lastDotIndex = moduleType.rfind('.');
+  if (lastDotIndex != std::string::npos) {
+    moduleType = moduleType.substr(lastDotIndex + 1);
+  }
+  return prefix + "(" + moduleType + ")";
+}
+
 void ReconstructScopesPass::constructRelativeNamesForModules(
     const Module& module,
     const std::string& prefix) {
-  module_names[module._ivalue()] = prefix;
+  module_names[module._ivalue()] = getModuleName(module, prefix);
   for (const NameModule& s : module.named_children()) {
-    constructRelativeNamesForModules(s.value, prefix + "." + s.name);
+    constructRelativeNamesForModules(
+        s.value, module_names[module._ivalue()] + "." + s.name);
+  }
+}
+
+void ReconstructScopesPass::appendSourceRangeInfo(
+    std::string& scopeString,
+    const InlinedCallStackEntry& frame) const {
+  SourceRange r = frame.second;
+  if (r.source()) {
+    if (auto orig = r.source()->findSourceRangeThatGenerated(r)) {
+      r = *orig;
+    }
+  }
+  if (auto file_line_col = r.file_line_col()) {
+    std::string filename;
+    size_t line, col;
+    std::tie(filename, line, col) = *file_line_col;
+    scopeString += "<" + filename + ":" + c10::to_string(line) + ":" +
+        c10::to_string(col) + ">";
   }
 }
 
-std::string ReconstructScopesPass::getScopeString(Function* f) const {
+std::string ReconstructScopesPass::getScopeString(
+    const InlinedCallStackEntry& frame) const {
+  Function* f = frame.first;
   if (!func_to_module.count(f)) {
     return "<null (no func in the map)>";
   }
   auto m = func_to_module.at(f);
   if (!module_names.count(m)) {
     return "<null (no module in the map)>";
   }
-  return module_names.at(m) + "." + f->name();
+  std::string scopeString = module_names.at(m) + "." + f->name();
+
+  if (class_types_are_not_unique_) {
```

Comment:
Put a comment here describing why we do this only if class types are not unique. Mainly to provide a little more information in the debug.